File: orion-platform-service/src/repositories/SecretRepository.ts
/**
 * SecretRepository - 数据访问层
 *
 * 负责 secrets 表的 CRUD 操作。
 * encrypted_value 以 Buffer 形式存储（PostgreSQL BYTEA），
 * 加密/解密逻辑在 SecretsService 层处理。
 */

import { BaseRepository } from '../db/base-repository';

export type SecretScope = 'org' | 'project' | 'environment';

export interface SecretEntity {
  id: string;
  tenantId: string;
  name: string;
  encryptedValue: Buffer;
  scope: SecretScope;
  createdAt: Date;
  updatedAt: Date;
  createdBy?: string;
}

export interface SecretCreateInput {
  id?: string;
  tenantId: string;
  name: string;
  encryptedValue: Buffer;
  scope?: SecretScope;
  createdBy?: string;
}

export class SecretRepository extends BaseRepository<SecretEntity> {
  constructor(db: { query: (text: string, params?: unknown[]) => Promise<{ rows: any[]; rowCount: number | null }> }) {
    super(db, 'secrets');
  }

  /**
   * 根据 tenant + name + scope 查找 secret
   */
  async findByTenantAndName(tenantId: string, name: string, scope?: SecretScope): Promise<SecretEntity | undefined> {
    if (scope) {
      const result = await this.db.query(
        `SELECT * FROM ${this.tableName} WHERE tenant_id = $1 AND name = $2 AND scope = $3`,
        [tenantId, name, scope],
      );
      if (result.rows.length === 0) return undefined;
      return this.mapRowToEntity(result.rows[0]);
    }

    // Without scope, search across all scopes (priority: project > environment > org)
    const result = await this.db.query(
      `SELECT * FROM ${this.tableName} WHERE tenant_id = $1 AND name = $2 ORDER BY
        CASE scope WHEN 'project' THEN 1 WHEN 'environment' THEN 2 WHEN 'org' THEN 3 END
        LIMIT 1`,
      [tenantId, name],
    );
    if (result.rows.length === 0) return undefined;
    return this.mapRowToEntity(result.rows[0]);
  }

  /**
   * 列出某个 tenant + scope 下的所有 secrets
   */
  async listByTenantAndScope(tenantId: string, scope?: SecretScope): Promise<SecretEntity[]> {
    let query: string;
    let params: unknown[];

    if (scope) {
      query = `SELECT id, tenant_id, name, scope, created_at, updated_at, created_by FROM ${this.tableName} WHERE tenant_id = $1 AND scope = $2 ORDER BY name ASC`;
      params = [tenantId, scope];
    } else {
      query = `SELECT id, tenant_id, name, scope, created_at, updated_at, created_by FROM ${this.tableName} WHERE tenant_id = $1 ORDER BY scope, name ASC`;
      params = [tenantId];
    }

    const result = await this.db.query(query, params);
    return result.rows.map((row: any) => this.mapRowToEntity(row));
  }

  /**
   * 创建 secret（带 ON CONFLICT 处理）
   */
  async upsert(data: SecretCreateInput): Promise<SecretEntity> {
    const scope = data.scope || 'project';
    const query = `
      INSERT INTO ${this.tableName} (id, tenant_id, name, encrypted_value, scope, created_by)
      VALUES ($1, $2, $3, $4, $5, $6)
      ON CONFLICT (tenant_id, name, scope)
      DO UPDATE SET encrypted_value = $4, updated_at = NOW()
      RETURNING *
    `;
    const result = await this.db.query(query, [
      data.id || this.generateId(),
      data.tenantId,
      data.name,
      data.encryptedValue,
      scope,
      data.createdBy || null,
    ]);
    if (result.rows.length === 0) {
      throw new Error(`UPSERT on ${this.tableName} returned no rows`);
    }
    return this.mapRowToEntity(result.rows[0]);
  }

  /**
   * 按 tenant + scope 批量删除
   */
  async deleteByTenantAndScope(tenantId: string, scope: SecretScope): Promise<number> {
    const result = await this.db.query(
      `DELETE FROM ${this.tableName} WHERE tenant_id = $1 AND scope = $2`,
      [tenantId, scope],
    );
    return result.rowCount ?? 0;
  }

  /**
   * 按 tenant 删除所有 secrets
   */
  async deleteByTenant(tenantId: string): Promise<number> {
    const result = await this.db.query(
      `DELETE FROM ${this.tableName} WHERE tenant_id = $1`,
      [tenantId],
    );
    return result.rowCount ?? 0;
  }

  protected mapRowToEntity(row: any): SecretEntity {
    return {
      id: row.id,
      tenantId: row.tenant_id,
      name: row.name,
      encryptedValue: row.encrypted_value as Buffer,
      scope: row.scope as SecretScope,
      createdAt: row.created_at,
      updatedAt: row.updated_at,
      createdBy: row.created_by,
    };
  }

  private generateId(): string {
    // Use crypto.randomUUID if available, fallback to simple id
    try {
      return require('crypto').randomUUID();
    } catch {
      return `secret-${Date.now()}-${Math.random().toString(36).slice(2, 9)}`;
    }
  }
}

File: orion-secret-svc-go/internal/repository/secret_repository.go
package repository

import (
	"context"
	"fmt"

	"orion/secret-svc-go/internal/models"

	"github.com/jmoiron/sqlx"
)

// Repository provides data access for the secrets table.
type Repository struct {
	db *sqlx.DB
}

// NewRepository creates a new Repository.
func NewRepository(db *sqlx.DB) *Repository {
	return &Repository{db: db}
}

// Create inserts a new secret. Uses ON CONFLICT for upsert semantics.
func (r *Repository) Create(ctx context.Context, s *models.Secret) error {
	query := `
		INSERT INTO secrets (id, tenant_id, name, value_encrypted, scope, description, created_by, version, environment)
		VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
		ON CONFLICT (tenant_id, name, scope)
		DO UPDATE SET value_encrypted = $4, description = $6, updated_at = NOW()
	`
	_, err := r.db.ExecContext(ctx, query,
		s.ID, s.TenantID, s.Name, s.Value, s.Scope, s.Description, s.CreatedBy, s.Version, s.Env,
	)
	return err
}

// List returns paginated secrets for a tenant (value columns excluded from SELECT).
func (r *Repository) List(ctx context.Context, tenantID string, offset, limit int) ([]models.Secret, error) {
	var items []models.Secret
	query := `SELECT id, tenant_id, name, scope, description, created_by, version, environment, created_at, updated_at
		FROM secrets WHERE tenant_id=$1 ORDER BY created_at DESC OFFSET $2 LIMIT $3`
	err := r.db.SelectContext(ctx, &items, query, tenantID, offset, limit)
	return items, err
}

// ListByScope returns secrets filtered by tenant and scope.
func (r *Repository) ListByScope(ctx context.Context, tenantID string, scope models.SecretScope) ([]models.Secret, error) {
	var items []models.Secret
	query := `SELECT id, tenant_id, name, scope, description, created_by, version, environment, created_at, updated_at
		FROM secrets WHERE tenant_id=$1 AND scope=$2 ORDER BY name ASC`
	err := r.db.SelectContext(ctx, &items, query, tenantID, scope)
	return items, err
}

// GetByID returns a secret by tenant and ID, including the encrypted value.
func (r *Repository) GetByID(ctx context.Context, tenantID, id string) (*models.Secret, error) {
	var s models.Secret
	err := r.db.GetContext(ctx, &s, `SELECT * FROM secrets WHERE id=$1 AND tenant_id=$2`, id, tenantID)
	if err != nil {
		return nil, fmt.Errorf("secret not found: %w", err)
	}
	return &s, nil
}

// FindByID returns a secret by ID without tenant filter.
// DEPRECATED: Use GetByID with tenantID instead. Retained only for cross-service
// internal calls where tenant context is not available.
func (r *Repository) FindByID(ctx context.Context, id string) (*models.Secret, error) {
	var s models.Secret
	err := r.db.GetContext(ctx, &s, `SELECT * FROM secrets WHERE id=$1`, id)
	if err != nil {
		return nil, fmt.Errorf("secret not found: %w", err)
	}
	return &s, nil
}

// FindByTenantAndName finds a secret by tenant and name, optionally filtered by scope.
// When scope is empty, searches across all scopes with priority: project > environment > org.
func (r *Repository) FindByTenantAndName(ctx context.Context, tenantID, name string, scope models.SecretScope) (*models.Secret, error) {
	if scope != "" {
		var s models.Secret
		err := r.db.GetContext(ctx, &s,
			`SELECT * FROM secrets WHERE tenant_id=$1 AND name=$2 AND scope=$3`,
			tenantID, name, scope,
		)
		if err != nil {
			return nil, fmt.Errorf("secret not found: %w", err)
		}
		return &s, nil
	}

	// Without scope, search across all scopes with priority ordering
	var s models.Secret
	err := r.db.GetContext(ctx, &s,
		`SELECT * FROM secrets WHERE tenant_id=$1 AND name=$2
		 ORDER BY CASE scope WHEN 'project' THEN 1 WHEN 'environment' THEN 2 WHEN 'org' THEN 3 END
		 LIMIT 1`,
		tenantID, name,
	)
	if err != nil {
		return nil, fmt.Errorf("secret not found: %w", err)
	}
	return &s, nil
}

// Delete removes a secret by tenant and ID.
func (r *Repository) Delete(ctx context.Context, tenantID, id string) error {
	_, err := r.db.ExecContext(ctx, `DELETE FROM secrets WHERE id=$1 AND tenant_id=$2`, id, tenantID)
	return err
}

// DeleteByID removes a secret by ID only.
// DEPRECATED: Use Delete with tenantID instead. Retained only for cross-service
// internal calls where tenant context is not available.
func (r *Repository) DeleteByID(ctx context.Context, id string) error {
	_, err := r.db.ExecContext(ctx, `DELETE FROM secrets WHERE id=$1`, id)
	return err
}

// DeleteByTenantAndScope removes all secrets for a tenant and scope.
func (r *Repository) DeleteByTenantAndScope(ctx context.Context, tenantID string, scope models.SecretScope) (int64, error) {
	result, err := r.db.ExecContext(ctx, `DELETE FROM secrets WHERE tenant_id=$1 AND scope=$2`, tenantID, scope)
	if err != nil {
		return 0, err
	}
	return result.RowsAffected()
}

// Count returns the total number of secrets for a tenant.
func (r *Repository) Count(ctx context.Context, tenantID string) (int, error) {
	var count int
	err := r.db.GetContext(ctx, &count, `SELECT COUNT(*) FROM secrets WHERE tenant_id=$1`, tenantID)
	return count, err
}

// UpdateValue updates the encrypted value of a secret, scoped to tenant.
func (r *Repository) UpdateValue(ctx context.Context, tenantID, id, encryptedValue string) error {
	_, err := r.db.ExecContext(ctx,
		`UPDATE secrets SET value_encrypted=$1, updated_at=NOW() WHERE id=$2 AND tenant_id=$3`,
		encryptedValue, id, tenantID,
	)
	return err
}

// UpdateDescription updates the description of a secret, scoped to tenant.
func (r *Repository) UpdateDescription(ctx context.Context, tenantID, id string, description *string) error {
	_, err := r.db.ExecContext(ctx,
		`UPDATE secrets SET description=$1, updated_at=NOW() WHERE id=$2 AND tenant_id=$3`,
		description, id, tenantID,
	)
	return err
}
